parlai/core/pytorch_data_teacher.py

```diff
@@ -0,0 +1,306 @@
+            self.dataloader = DataLoader(
+                self.dataset,
+                batch_size=self.bsz,
+                shuffle=False,
+                sampler=sampler.SequentialSampler(self.dataset),
+                num_workers=self.num_workers,
+                collate_fn=collate_fn,
+                pin_memory=False,
+                drop_last=False,
+                timeout=0)
+            self.lastYs = [None] * self.bsz
+        else:
+            self.dataset = shared['dataset']
+            self.dataloader = shared['dataloader']
+            self.lastYs = shared['lastYs']
+
+        self.num_batches = math.ceil(self.dataset.num_examples()/self.bsz)
+        self.reset()
+
+    def reset(self):
+        """Reset the dialog so that it is at the start of the epoch,
+        and all metrics are reset.
+        """
+        super().reset()
+        self.metrics.clear()
+        self.reset_data()
+
+    def reset_data(self):
+        self.data = enumerate(self.dataloader)
+        self.lastY = None
+        self.epochDone = False
+        self.episode = None
+        self.episode_done = True
+        self.episode_idx = 0
+        self.data = enumerate(self.dataloader)
+
+    def share(self):
+        shared = super().share()
+        shared['dataloader'] = self.dataloader
+        shared['dataset'] = self.dataset
+        shared['lastYs'] = self.lastYs
+        return shared
+
+    def next_example(self):
+        if self.episode_done:
+            try:
+                self.episode_idx, self.episode = next(self.data)
+                self.entry_idx = 0
+                epoch_done = False
+            except StopIteration:
+                ex = {'episode_done': True, 'id': self.getID()}
+                epoch_done = True
+        else:
+            self.entry_idx += 1
+
+        if not epoch_done:
+            ex = self.episode[self.entry_idx]
+            self.episode_done = ex['episode_done']
+            if (self.episode_done
+                    and self.episode_idx + self.step_size >= self.num_episodes()):
+                epoch_done = True
+
+        return ex, epoch_done
+
+    def next_batch(self):
+        try:
+            batch_idx, batch = next(self.data)
+            epoch_done = False
+        except StopIteration:
+            batch = [{'episode_done': True, 'id': self.getID()}] * self.bsz
+            epoch_done = True
+        if not epoch_done and batch_idx == self.num_batches:
+            epoch_done = True
+
+        return batch, epoch_done
+
+    def num_episodes(self):
+        """Get the number of episodes in this dataset."""
+        return self.dataset.num_episodes()
+
+    def num_examples(self):
+        """Get the total number of examples in this dataset."""
+        return self.dataset.num_examples()
+
+    def observe(self, observation):
+        """Process observation for metrics."""
+        if self.bsz > 1:
+            self.lastY = self.lastYs[self.batchindex]
```

Comment:
also need `self.lastYs[self.batchindex] = None` (had a few changes in FixedDialogTeacher, this might not be the only one)